item toggling giga rework + full ninja refactor

deltanedas opened the pull request:
## About the PR
depends on #28208 and #28216

reworked some existing item toggle components:
- made everything use `Entity<T>`
- fleshed out its api a bit more and made it more robust
- you can have itemtoggles that dont toggle when pressing z, `onUse: false`
- added ComponentToggler to add and remove components whenever it gets toggled
- removed ItemToggleSharp and ItemToggleDisarmMalus, outclassed by ComponentToggler
- due to above, system is now only in shared
- fixed unpredictable toggles using wrong functions to play sounds

added ToggleClothingComponent to give clothing an action to toggle itself, or the user (valid cat ears added to play with this)

a few things changed to now use ItemToggle:
- instead of a second entirely copy pasted component for speed modifying it just has speed modifying work with itemtoggle
- magboots, which now also just use 1 action instead of 1 for every colour /!\
- speedboots
- anything that used powerdraw, most of them now inherit BaseHandheldComputer for a little less copypasta
- powercelldraw uses itemtoggle instead of Drawing, which is replaced by Enabled for borgs usecase. anything else is probably fine just using itemtoggle in yaml
- borg uses it because above with disabling power usage when theres no mind
- defibs use it for logic now
- ProximityDetector and Beeper use ItemToggle, IsMuted is now actually used to mute it when theres nothing found crazy idea
- Reflect using ItemToggle for logic + fixed wrongly removing alert when unequipping 1 of multiple reflective items + removed 0-prob reflect from base species mob.

refactored all ninja code:
- made it use ItemToggle for suit stealth and gloves
- stealth clothing replaced with just adding StealthComponent with ComponentToggler
- gloves have abilities defined in yaml with ComponentRegistry my beloved
- moved throwing star ability into ItemCreatorComponent
- made spawning actually work now with antag refactor

misc refactor:
- gravity no longer has magboots hardcoded, inv relay GravityCheckEvent added
- small charges system refactor and added TryUseCharges for easy non-emag usecases

TODO:
- make borg use movespeedmodifier somehow?
- test tes te test est tes testst

in the future probably:
- fire extinguisher using ItemToggle and ToggleVerb to control safety
- idk more

## Why / Balance
the shitcode has aged like fine wine now its time to bottle it

## Media
- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
- PowerCellDraw, Magboots, ClothingSpeedModifier now all use ItemToggle
- all magboots use the same ActionToggleMagboots instead of individual actions copy pasted
- SharedItemToggleSystem renamed to ItemToggleSystem and uses `Entity<T>` in the api
- removed ItemToggleDisarmMalus and ItemToggleSharp components, use ComponentToggler instead


**Changelog**
fake cl:
- sentient magboots keep themselves on the floor when theres no gravity

AJCM-git reviewed (changes requested):
Your best pr yet soldier (tho in an ideal world the ninja changes would be another pr requiring this one smh).
Now from just testing and reviewing the yaml-side:
- The only thing in the TODO that really needs doing here is the fire extinguisher item toggle
- The toggle action on clothing and verb need more work, it should probably only be added if they are on the correct slot and deactivate when they are not, OR ensure it deactivates when not in hand (the valid cat ears leave you with the kill sign when on the ground, but we still want to enable someone else's magboots and drop them on the ground for them to pick up, so maybe a bool somewhere???)
- Blood red magboots dont have the toggle magboots action

AJCM-git commented on Content.Shared/Item/ItemToggle/ComponentTogglerSystem.cs:
```diff
@@ -0,0 +1,28 @@
+    private void OnToggled(Entity<ComponentTogglerComponent> ent, ref ItemToggledEvent args)
+    {
+        var target = ent.Comp.Parent ? Transform(ent).ParentUid : ent.Owner;
+        if (target == null)
```
target cant be null

AJCM-git commented on Content.Shared/Item/ItemToggle/ItemToggleSystem.cs:
```diff
@@ -105,50 +125,42 @@ public bool TryActivate(EntityUid uid, EntityUid? user = null, bool predicted =
-    private void Activate(EntityUid uid, ItemToggleComponent itemToggle, bool predicted, EntityUid? user = null)
+    private void Activate(Entity<ItemToggleComponent> ent, bool predicted, EntityUid? user = null)
     {
-        // TODO: Fix this hardcoding
```
delete the         // END FIX HARDCODING too

deltanedas commented:
> if they are on the correct slot and deactivate when they are not

i think the **clothing** got equipped event accounts for this

> Blood red magboots dont have the toggle magboots action

valid + ratio

AJCM-git commented on Content.Shared/Item/ItemToggle/ItemToggleSystem.cs:
```diff
@@ -204,55 +217,56 @@ private void UpdateVisuals(Entity<ItemToggleComponent> ent)
     {
-        if (!itemToggle.Activated)
-            TryActivate(uid, itemToggle: itemToggle);
+        // FIXME: for some reason both client and server play sound
```
dew it, you prob messed up updating comp.Predictable (which seems like an unnecesary field to have tbh)

deltanedas replied:
welders need to disable prediction and i couldnt fix the wield thing mispredicting dont know why

AJCM-git commented on Content.Shared/Clothing/SharedMagbootsSystem.cs:
```diff
@@ -1,101 +1,78 @@
-
-        if (TryComp<ItemComponent>(uid, out var item))
+        // only stick to the floor if being worn in the correct slot
+        if (_container.TryGetContainingContainer(ent, out var container) &&
```
obsolete

AJCM-git commented on Content.Shared/Clothing/ClothingSpeedModifierSystem.cs:
```diff
@@ -142,60 +110,15 @@ private void OnClothingVerbExamine(EntityUid uid, ClothingSpeedModifierComponent
-
-        var user = args.User;
-        ActivationVerb verb = new()
+        if (_container.TryGetContainingContainer(ent, out var container))
```
a

AJCM-git commented on Content.Shared/Beeper/Components/BeeperComponent.cs:
```diff
@@ -13,12 +13,6 @@ namespace Content.Shared.Beeper.Components;
 [RegisterComponent, NetworkedComponent, Access(typeof(BeeperSystem)), AutoGenerateComponentState]
```
requires item toggling now

AJCM-git reviewed (changes requested):
Remember checking for unused namespaces in general, also you broke gravity (you really should only touch whats absolutely necessary for toggling)
Also pretty much every new Obsolete is unnecesary so you can just remove the method (and update the ones that arent)

AJCM-git commented:
> i think the clothing got equipped event accounts for this

Does it work like this: Toggle magboots on (either verb or action) > Equip them and have the effect > Unequip them and no longer have the effect (They are still activated) > Equip them and have the effect

Cuz the valid ears didnt work like this

deltanedas commented on Content.Shared/Ninja/Systems/SharedNinjaSuitSystem.cs:
```diff
@@ -14,137 +17,159 @@ namespace Content.Shared.Ninja.Systems;
-        _actionContainer.EnsureAction(uid, ref component.EmpActionEntity, component.EmpAction);
-        Dirty(uid, component);
+        var user = args.Wearer;
+        if (_ninja.NinjaQuery.TryComp(user, out var ninja));
```
always execute the NinjaEquipped bit :trollface:

deltanedas commented on Content.Shared/Ninja/Systems/DashAbilitySystem.cs:
```diff
@@ -16,55 +16,48 @@ namespace Content.Shared.Ninja.Systems;
 {
+    [Dependency] private readonly ActionContainerSystem _actionContainer = default!;
     [Dependency] private readonly IGameTiming _timing = default!;
     [Dependency] private readonly SharedAudioSystem _audio = default!;
```
it uses EnsureAction this is not unused

AJCM-git replied:
IIRC i was refering to the audio one

AJCM-git replied:
and MAYBE timing

AJCM-git replied:
also remember to re-request reviews when you have cant continue (cuz of waiting responses in one or resolving others all others) nerd

deltanedas replied:
they are both used but actioncontainer somehow got put at the bottom so fixed that

deltanedas commented:
> Does it work like this: Toggle magboots on (either verb or action) > Equip them and have the effect > Unequip them and no longer have the effect (They are still activated) > Equip them and have the effect

the staying on after taking them off is, i just fixed a bug where if you toggle it the speed you get is the previous state, so turn magboots on to be fast and off to be slow (opposite of how it should be)

this is now fixed so thats probably what you meant?

deltanedas commented:
thats a separate thing where unequipped is raised after its already gone :trollface: itll just have to actually track the wearer

metalgearsloth commented:
is there a contrib notif for this.

deltanedas commented:
@AJCM-git pro review
